src/System.Net.Http/src/System/Net/Http/HPack/HPackDecoder.cs

```diff
@@ -0,0 +1,398 @@
+        private const byte LiteralHeaderFieldWithoutIndexingRepresentation = 0x00;
+
+        // http://httpwg.org/specs/rfc7541.html#rfc.section.6.2.3
+        //   0   1   2   3   4   5   6   7
+        // +---+---+---+---+---+---+---+---+
+        // | 0 | 0 | 0 | 1 |  Index (4+)   |
+        // +---+---+-----------------------+
+        private const byte LiteralHeaderFieldNeverIndexedMask = 0xf0;
+        private const byte LiteralHeaderFieldNeverIndexedRepresentation = 0x10;
+
+        // http://httpwg.org/specs/rfc7541.html#rfc.section.6.3
+        //   0   1   2   3   4   5   6   7
+        // +---+---+---+---+---+---+---+---+
+        // | 0 | 0 | 1 |   Max size (5+)   |
+        // +---+---------------------------+
+        private const byte DynamicTableSizeUpdateMask = 0xe0;
+        private const byte DynamicTableSizeUpdateRepresentation = 0x20;
+
+        // http://httpwg.org/specs/rfc7541.html#rfc.section.5.2
+        //   0   1   2   3   4   5   6   7
+        // +---+---+---+---+---+---+---+---+
+        // | H |    String Length (7+)     |
+        // +---+---------------------------+
+        private const byte HuffmanMask = 0x80;
+
+        private const int IndexedHeaderFieldPrefix = 7;
+        private const int LiteralHeaderFieldWithIncrementalIndexingPrefix = 6;
+        private const int LiteralHeaderFieldWithoutIndexingPrefix = 4;
+        private const int LiteralHeaderFieldNeverIndexedPrefix = 4;
+        private const int DynamicTableSizeUpdatePrefix = 5;
+        private const int StringLengthPrefix = 7;
+
+        private readonly int _maxDynamicTableSize;
+        private readonly DynamicTable _dynamicTable;
+        private readonly IntegerDecoder _integerDecoder = new IntegerDecoder();
+        private readonly byte[] _stringOctets = new byte[MaxStringOctets];
+        private readonly byte[] _headerNameOctets = new byte[MaxStringOctets];
+        private readonly byte[] _headerValueOctets = new byte[MaxStringOctets];
+
+        private State _state = State.Ready;
+        private byte[] _headerName;
+        private int _stringIndex;
+        private int _stringLength;
+        private int _headerNameLength;
+        private int _headerValueLength;
+        private bool _index;
+        private bool _huffman;
+
+        public HPackDecoder(int maxDynamicTableSize = DefaultHeaderTableSize)
+            : this(maxDynamicTableSize, new DynamicTable(maxDynamicTableSize))
+        {
+            _maxDynamicTableSize = maxDynamicTableSize;
+        }
+
+        // For testing.
+        internal HPackDecoder(int maxDynamicTableSize, DynamicTable dynamicTable)
+        {
+            _maxDynamicTableSize = maxDynamicTableSize;
+            _dynamicTable = dynamicTable;
+        }
+
+        public void Decode(ReadOnlySpan<byte> data, HeaderCallback onHeader)
+        {
+            for (var i = 0; i < data.Length; i++)
+            {
+                OnByte(data[i], onHeader);
```

Comment:
We call this function for every byte in the received header data?  Seems like a potentially significant expense :)

Reply:
I agree. Generally I've avoided making changes to this code, since it came from ASP.NET and I'm not sure what our code sharing plans are. That said, I think we should just make whatever changes we feel are appropriate. I'll tackle this.